### PR TITLE
Migrate docker-release (Build Official Docker Images) to OSDC

### PR DESCRIPTION
**Pull request by huydhn**
Migrates docker-release (Build Official Docker Images) off the EC2 `linux.2xlarge` runner onto OSDC + remote BuildKit, mirroring `build-manywheel-images`.

- **generate-matrix** (`ubuntu-24.04`): ref-gates the build runner — release-isolated `rel-l-*` on `main`/`release/*`/`v*`, regular `l-*` otherwise — and bakes a per-platform `runner` into each matrix entry.
- **build** (OSDC `actions-runner` container → remote BuildKit `tcp://buildkitd-{amd64,arm64}`): still builds via `make -f docker.Makefile` (`USE_BUILDX=1`, so its buildx targets the remote builder); `make`/Python come from `apt` + `setup-uv`. Each single-platform build is native to its runner arch, so QEMU is dropped.
- **Push nightly tags**: unchanged behavior, run daemon-less — reads `torch.version.git_version` from the pushed image via a one-file BuildKit `--output` build, then creates the `<commit>-cu<cuda>` / `latest` aliases with `docker buildx imagetools create`.
- Dropped host-only steps (ARC has no docker daemon): setup-ssh / setup-linux / teardown / QEMU / ecr-login (base is public `ubuntu:24.04`).

**DONE:** add `docker-release.yml@refs/heads/nightly` and `@refs/tags/v*` to the release runner group's selected-workflows allowlist, or protected-ref builds hang queued.

Authored with assistance from Claude Code.

### CODE REVIEW

**Comment by atalman** [comment]
### Review: can this break ghcr.io? — no critical issues

Reviewed with a focus on the published `ghcr.io/pytorch/pytorch-nightly` images. **No image-corrupting, mis-targeting, or security issues** — the four ways this could have broken ghcr are all handled:

- Registry target is set explicitly (`DOCKER_REGISTRY: ghcr.io`, `DOCKER_ORG: pytorch`, `DOCKER_IMAGE` exported per-ref), so the Makefile's daemon-less `DOCKER_ORG ?= $(shell docker info … Username)` fallback (empty in the ARC container) is never hit.
- `WITH_PUSH` still gates login + `--push` to `push` on `refs/heads/nightly` / `refs/tags/v*` only → no pushes from `pull_request` / `workflow_dispatch`.
- Per-arch tags are distinct (amd64 `…-cudaX-cudnnY-{runtime,devel}`, arm64 `…-runtime` CPU) → the amd64/arm64 split can't clobber a shared tag.
- `imagetools create` is registry-side; `latest` is created only from the stable-CUDA amd64 image, and a failed alias leaves the prior tag intact.

#### Highest-impact item (operational, please confirm before merge)
The runner-group **selected-workflows allowlist** noted in the description is a hard gate, not just a TODO: if `docker-release.yml@refs/heads/nightly` (and `@refs/tags/v*`) isn't added before this lands, protected-ref builds hang queued and **nightly ghcr tags silently stop updating** (stale images). Please sequence that with the merge.

#### Non-blocking, worth fixing
1. **Retry loop can retry a doomed build for ~90 min.** The classifier greps the *entire* `make`/buildx output (`run_build 2>&1 | tee "$build_log"`) for generic substrings (`EOF`, `i/o timeout`, `connection reset`, `transport: Error`, `no such host`). A real build failure whose logs contain any of these (an apt/pip network blip, a compiler error) is misread as "BuildKit not ready" and re-runs the full build up to `360 × 15s`, burning the 240-min timeout and delaying the failure signal. Scope the match to the connect phase (or only when no build step has started), and/or lower `attempts`.
2. **`PREFIX="ghcr.io/pytorch/pytorch-nightly"` is hardcoded** in *Push nightly tags*, duplicating the `DOCKER_REGISTRY`/`DOCKER_ORG`/`DOCKER_IMAGE` env used by the build — two sources of truth. Use `${DOCKER_REGISTRY}/${DOCKER_ORG}/${DOCKER_IMAGE}`.
3. **`container: image: ghcr.io/actions/actions-runner:latest` is unpinned**, unlike the SHA-pinned actions elsewhere — a future push to that tag can change tooling/shell assumptions and break the job nondeterministically. Pin to a digest.
4. **git_version extraction pulls the full multi-GB image to read a 7-char hash** and makes `latest`/`<commit>` aliasing depend on `import torch` at build time (a transient import failure leaves a partial publish: `…-runtime` pushed, `<commit>-cu*`/`latest` missing). Prefer stamping git_version as an image LABEL and reading it via `docker buildx imagetools inspect --format` (no pull, no re-run of `import torch`).

Net: safe to merge from a "will it break ghcr" standpoint provided the allowlist is in place; (1) and (3) are the ones I'd fix before relying on it in production.

_Reviewed with assistance from Claude Code._

File: .github/scripts/generate_docker_release_matrix.py [hunk 1]
```diff
@@ -20,6 +21,11 @@
 
 
 def generate_docker_matrix() -> dict[str, list[dict[str, str]]]:
+    # OSDC runner per platform; the workflow ref-gates these (release-isolated
+    # on protected refs, regular otherwise) and passes them in.
+    x86_runner = os.environ.get("RUNNER_X86", "mt-l-x86iavx512-48-384")
```
Comment (jeanschmidt) [inline .github/scripts/generate_docker_release_matrix.py:26]: I don't like how hardcoded and hard to find this feels

File: .github/workflows/docker-release.yml [hunk 2]
```diff
@@ -50,15 +50,30 @@ jobs:
       - name: Get docker release matrix
         id: generate-matrix
         run: |
+          # Release-isolated OSDC runners on protected refs, regular otherwise.
+          if [[ "${GITHUB_REF}" == refs/heads/main || "${GITHUB_REF}" == refs/heads/nightly || "${GITHUB_REF}" == refs/heads/release/* || "${GITHUB_REF}" == refs/tags/v* ]]; then
+            export RUNNER_X86="mt-rel-l-x86iavx512-44-340"
```
Comment (jeanschmidt) [inline .github/workflows/docker-release.yml:55]: ideally this should be a single source of truth

File: .github/workflows/docker-release.yml [hunk 3]
```diff
@@ -130,33 +136,77 @@ jobs:
             echo "TRITON_VERSION=$(cut -f 1 .ci/docker/triton_version.txt)+$(cut -c -10 .ci/docker/ci_commit_pins/triton.txt)";
           } >> "${GITHUB_ENV}"
       - name: Run docker build / push
-        # WITH_PUSH is used here to determine whether or not to add the --push flag
+        # docker.Makefile drives the buildx build; the remote BuildKit builder
+        # set up above is the current buildx builder, so it runs there.
         run: |
-          make -f docker.Makefile "${BUILD_IMAGE_TYPE}-image"
+          set -eo pipefail
```
Comment (jeanschmidt) [inline .github/workflows/docker-release.yml:142]: I would assume that this is a script used in multiple places where we do docker builds

Maybe we should re-use it, as this is a common code pattern

File: .github/workflows/docker-release.yml (hunk 2 above)
Comment (huydhn) [reply .github/workflows/docker-release.yml:55]: This would probably call for a reusable workflow dealing with release runners. Let me try it out

File: .github/workflows/docker-release.yml (hunk 3 above)
Comment (huydhn) [reply .github/workflows/docker-release.yml:142]: The new stack is at https://github.com/pytorch/pytorch/pull/190619, which create a new reusable workflows to select the release runner plus add a script to make sure all Docker builds wait for Buildkit nodes in the same way

File: .github/scripts/generate_docker_release_matrix.py (hunk 1 above)
Comment (huydhn) [reply .github/scripts/generate_docker_release_matrix.py:26]: With https://github.com/pytorch/pytorch/pull/190606, we will only need to name the runner in one place, a new `.github/workflows/_select-release-runner.yml` reusable workflow

**Comment by huydhn** [comment]
Done in https://github.com/pytorch/pytorch/pull/190619
